Toggl.Foundation.MvvmCross/Collections/Diffing.cs

```diff
@@ -0,0 +1,767 @@
+
+        private List<SectionAssociatedData> initialSectionData;
+        private List<SectionAssociatedData> finalSectionData;
+
+        private List<List<ItemAssociatedData>> initialItemData;
+        private List<List<ItemAssociatedData>> finalItemData;
+
+        public Diffing(IEnumerable<TSection> initialSections, IEnumerable<TSection> finalSections)
+        {
+            this.initialSections = initialSections.ToList();
+            this.finalSections = finalSections.ToList();
+        }
+
+        public List<Changeset> computeDifferences()
+        {
+            (initialSectionData, finalSectionData) = calculateSectionMovements(initialSections.ToList(), finalSections.ToList());
+
+            initialItemCache = initialSections.Select(collection => collection.Items.ToList()).ToList();
+            finalItemCache = finalSections.Select(collection => collection.Items.ToList()).ToList();
+
+            var result = Enumerable.Empty<Changeset>().ToList();
+
+            (initialItemData, finalItemData) = calculateItemMovements(
+                initialItemCache,
+                finalItemCache,
+                initialSectionData,
+                finalSectionData
+            );
+
+            result.AddRange(generateDeleteSectionsDeletedItemsAndUpdatedItems());
+            result.AddRange(generateInsertAndMoveSections());
+            result.AddRange(generateInsertAndMovedItems());
+
+            return result;
+        }
+
+        private static (List<SectionAssociatedData>, List<SectionAssociatedData>) calculateSectionMovements(
+            List<TSection> initialSections, List<TSection> finalSections)
+        {
+            var initialSectionIndexes = indexSections(initialSections);
+
+            var initialSectionData = Enumerable.Range(0, initialSections.Count)
+                .Select(_ => SectionAssociatedData.Initial()).ToList();
+            var finalSectionData = Enumerable.Range(0, finalSections.Count)
+                .Select(_ => SectionAssociatedData.Initial()).ToList();
+
+            for (var i = 0; i < finalSections.Count; i++)
+            {
+                var section = finalSections[i];
+
+                finalSectionData[i].ItemCount = finalSections[i].Items.Count;
+
+                if (!initialSectionIndexes.ContainsKey(section.Identity))
+                {
+                    continue;
+                }
+                var initialSectionIndex = initialSectionIndexes[section.Identity];
+
+                if (initialSectionData[initialSectionIndex].MoveIndex.HasValue)
+                {
+                    throw new DuplicateSectionException(section.Identity);
+                }
+
+                initialSectionData[initialSectionIndex].MoveIndex = i;
+                finalSectionData[i].MoveIndex = initialSectionIndex;
+            }
+
+            var sectionIndexAfterDelete = 0;
+
+            // deleted sections
+
+            for (var i = 0; i < initialSectionData.Count(); i++)
```

Comment:
It's a list, use the Count property.

```suggestion
            for (var i = 0; i < initialSectionData.Count; i++)
```